assets/character/scripts/player_character.gd

```diff
@@ -74,6 +45,79 @@ func _physics_process(delta):
 		velocity = target_velocity
 	super._physics_process(delta)
 
+func on_item_selected(item: String):
+	selection_cursor.visible = false
+	if item.is_empty():
+		return
+	var item_row: ItemConfig = items_dt.get_row(item)
+	match(item_row.action_type):
+		ItemConfig.ActionType.PLANT:
+			var crop_details = GameManager.crops_dt.get_row(GameManager.selected_item)
+			selection_cursor.cell_select_predicate = plant_action_predicate
+			selection_cursor.visible = true
+			selection_cursor.mesh = load(crop_details.mesh) if crop_details.mesh else SphereMesh.new()
+			selection_cursor.radius = crop_details.effect_radius
+			return
+		ItemConfig.ActionType.WATER:
+			selection_cursor.cell_select_predicate = Callable()
+			selection_cursor.visible = true
+			selection_cursor.mesh = null
+			selection_cursor.radius = 5 # TODO: Make tied to upgrades for watering can
+			selection_cursor.selected_grid_prop = "hydration"
+			return
+		ItemConfig.ActionType.SCAN:
+			selection_cursor.cell_select_predicate = Callable()
+			selection_cursor.visible = true
+			selection_cursor.mesh = null
+			selection_cursor.radius = 5 # TODO: Make tied to upgrades for scanner
+			selection_cursor.selected_grid_prop = GameManager.scanner_prop
+			return
+
+func on_prop_updated(prop: String):
+	if GameManager.selected_item == 'scanner':
+		selection_cursor.selected_grid_prop = prop
+
+func on_mouse_down():
+	mouse_down = true
+	if GameManager.selected_item.is_empty():
+		var pos = Utils.get_perspective_collision_ray_point(self)
+		if pos:
+			navigate_to(pos)
+	elif selection_cursor and selection_cursor.visible: 
+		var cell = selection_cursor.hovered_cell
+		if cell and GameManager.current_zone.grid.is_cell_valid(cell):
+			if not selection_cursor.cell_select_predicate.is_valid() or selection_cursor.cell_select_predicate.call(cell):
+				start_selected_action()
+
+func start_selected_action():
+	var item_row: ItemConfig = items_dt.get_row(GameManager.selected_item)
+	match(item_row.action_type):
+		ItemConfig.ActionType.PLANT:
+			run_action(CharacterActionPlantCrop.new(self, selection_cursor.hovered_cell, GameManager.selected_item))
+			return
+		ItemConfig.ActionType.WATER:
+			run_action(CharacterActionWaterSoil.new(self, selection_cursor.hovered_cell))
+			return
+
+func on_mouse_up():
+	mouse_down = false
+	if current_action is CharacterActionWaterSoil:
+		current_action.complete()
```

Comment:
This really needs a big cleanup where we push all this responsibility into the actions